### PR TITLE
Qunit initial

### PR DESCRIPTION
**Pull request by WrathfulSpatula**
The bugged two-bit "Swap" was replaced with the length=1 case of the register version of "Swap," which is correct. Additionally, permutation initialization was corrected in QUnit, "auto" loop references were corrected for shard mutation, and "Decompose" was changed slightly to give a large optimization.

Passes all unit tests.

### CODE REVIEW

File: src/qengine/state/gates.cpp [hunk 1]
```diff
@@ -142,20 +142,17 @@ void QEngineCPU::X(bitLenInt start, bitLenInt length)
...
-    int distance = start1 - start2;
-    if (distance < 0) {
+    else if (distance < 0) {
         distance *= -1;
     }
     if (distance < length) {
         bitLenInt i;
         for (i = 0; i < length; i++) {
-            Swap(start1 + i, start2 + i);
+            Swap(start1 + i, start2 + i, 1);
```
Comment (bennbollay) [inline src/qengine/state/gates.cpp:155]: Check if this infinitely loops

File: test/tests.cpp [hunk 2]
```diff
@@ -593,25 +594,26 @@ TEST_CASE_METHOD(QInterfaceTestFixture, "test_entanglement")
     REQUIRE_THAT(qftReg, HasProbability(0, 20, 0x1));
 }
 
+/*
```
Comment (bennbollay) [inline test/tests.cpp:597]: FYI #if 0 is an easier way.

**Review by bennbollay — APPROVED** [review]
As long as that swap code path actually looks good, merge.

File: src/qengine/state/gates.cpp (hunk 1 above)
Comment (WrathfulSpatula) [reply src/qengine/state/gates.cpp:155]: It does not. Done.

File: test/tests.cpp (hunk 2 above)
Comment (WrathfulSpatula) [reply test/tests.cpp:597]: Thank you, noted.

**Pull request merged**